File: src/state.py
"""Storage paths and state persistence for gh-readme-pipeline.

Addresses: H1 (resume), H7 (summary aggregation), M2 (XDG paths).

Public API:
    xdg_cache_dir() -> Path
    xdg_state_dir() -> Path
    StateStore(user, state_dir=None)
        .record(repo_name, status, mode=None, error=None, pr_url=None)
        .load_processed() -> set[str]
        .summary() -> dict
    prompt_resume(processed_count: int) -> str
"""
import json
import os
import re
import threading
from datetime import datetime, timezone
from pathlib import Path

# RT-L1: GitHub username regex (1-39 chars, alnum or hyphen, no leading/trailing/double hyphen).
_GH_USER_RE = re.compile(r"^[A-Za-z0-9](?:[A-Za-z0-9]|-(?=[A-Za-z0-9])){0,38}$")


# ---------------------------------------------------------------------------
# XDG path helpers
# ---------------------------------------------------------------------------

APP_NAME = "gh-readme-pipeline"

_PROCESSED_STATUSES = frozenset({"pushed", "pr_opened", "commit_only"})


def xdg_cache_dir() -> Path:
    """Return the application cache directory, honoring XDG_CACHE_HOME.

    Falls back to ``~/.cache/gh-readme-pipeline`` if XDG_CACHE_HOME is unset.
    The directory is NOT created; callers are responsible for mkdir.
    """
    base = os.environ.get("XDG_CACHE_HOME")
    if base:
        return Path(base) / APP_NAME
    return Path.home() / ".cache" / APP_NAME


def xdg_state_dir() -> Path:
    """Return the application state directory, honoring XDG_STATE_HOME.

    Falls back to ``~/.local/state/gh-readme-pipeline`` if XDG_STATE_HOME
    is unset.  The directory is NOT created; callers are responsible for mkdir.
    """
    base = os.environ.get("XDG_STATE_HOME")
    if base:
        return Path(base) / APP_NAME
    return Path.home() / ".local" / "state" / APP_NAME


# ---------------------------------------------------------------------------
# StateStore
# ---------------------------------------------------------------------------

class StateStore:
    """Append-only JSONL state file for tracking processed repositories.

    Each call to ``record()`` appends one line.  ``load_processed()`` and
    ``summary()`` scan the whole file on each call (file is ≤1000 lines).

    Args:
        user:       GitHub username — used as part of the state filename.
        state_dir:  Directory for the state file.  Defaults to
                    ``xdg_state_dir()``.  Useful for testing.
    """

    def __init__(self, user: str, state_dir: Path | None = None) -> None:
        # RT-L1: validate GH username — prevents path traversal via state filename.
        if not _GH_USER_RE.match(user):
            raise ValueError(f"invalid GitHub username: {user!r}")
        self._user = user
        self._state_dir = state_dir if state_dir is not None else xdg_state_dir()
        self._state_file = self._state_dir / f"state-{user}.jsonl"
        # P8: serialise writes within a process; parallel WorkerPool threads
        # call record() concurrently.
        self._lock = threading.Lock()

    # ------------------------------------------------------------------
    # Write
    # ------------------------------------------------------------------

    def record(
        self,
        repo_name: str,
        status: str,
        mode: str | None = None,
        error: str | None = None,
        pr_url: str | None = None,
    ) -> None:
        """Append one record to the state file.

        The record is a JSON object with at minimum ``repo``, ``status``, and
        ``ts`` fields.  Optional fields (``mode``, ``error``, ``pr_url``) are
        included only when non-None.

        The file is opened in append mode and flushed after each write to
        ensure atomicity on a single line.
        """
        entry: dict = {
            "repo": repo_name,
            "status": status,
            "ts": datetime.now(tz=timezone.utc).isoformat(timespec="seconds"),
        }
        if mode is not None:
            entry["mode"] = mode
        if error is not None:
            entry["error"] = error
        if pr_url is not None:
            entry["pr_url"] = pr_url

        self._state_dir.mkdir(parents=True, exist_ok=True)
        with self._lock:
            with self._state_file.open("a", encoding="utf-8") as fh:
                fh.write(json.dumps(entry) + "\n")
                fh.flush()

    # ------------------------------------------------------------------
    # Read
    # ------------------------------------------------------------------

    def has_prior_state(self) -> bool:
        """Return True if the state file exists on disk (CR-LOW-1)."""
        return self._state_file.exists()

    def _read_all(self) -> list[dict]:
        """Return all records from the state file; empty list if no file."""
        if not self._state_file.exists():
            return []
        records = []
        with self._state_file.open("r", encoding="utf-8") as fh:
            for line in fh:
                line = line.strip()
                if line:
                    try:
                        records.append(json.loads(line))
                    except json.JSONDecodeError:
                        pass  # skip malformed lines
        return records

    def load_processed(self) -> set[str]:
        """Return the set of repo names that completed successfully.

        A repo is considered processed if its last recorded status is one of:
        ``pushed``, ``pr_opened``, ``commit_only``.

        Duplicate records for the same repo are deduplicated (set semantics).
        """
        return {
            record["repo"]
            for record in self._read_all()
            if record.get("status") in _PROCESSED_STATUSES
        }

    def summary(self) -> dict:
        """Return an aggregation dict with:

        - Per-status counts (keys are status strings, values are int counts).
        - ``pr_urls``:      list of PR URLs for ``pr_opened`` records.
        - ``failed_repos``: list of repo names with ``failed`` status.
        """
        records = self._read_all()
        counts: dict[str, int] = {}
        pr_urls: list[str] = []
        failed_repos: list[str] = []

        for rec in records:
            status = rec.get("status", "unknown")
            counts[status] = counts.get(status, 0) + 1
            if status == "pr_opened" and rec.get("pr_url"):
                pr_urls.append(rec["pr_url"])
            if status == "failed":
                failed_repos.append(rec["repo"])

        return {
            **counts,
            "pr_urls": pr_urls,
            "failed_repos": failed_repos,
        }


# ---------------------------------------------------------------------------
# Resume prompt
# ---------------------------------------------------------------------------

_RESUME_CHOICES = {
    "r": "resume",
    "a": "all",
    "s": "fresh",
    "q": "quit",
}


def prompt_resume(processed_count: int) -> str:
    """Prompt the user for how to handle a prior state file.

    Displays the number of already-processed repos and asks:
        [r]esume (skip processed)
        [a]ll incl. failed
        [s]tart fresh
        [q]uit

    Re-prompts on invalid input until a valid choice is entered.

    Returns one of: ``"resume"``, ``"all"``, ``"fresh"``, ``"quit"``.
    """
    prompt_text = (
        f"Found {processed_count} repos already processed. "
        f"[r]esume (skip processed) / [a]ll incl. failed / "
        f"[s]tart fresh / [q]uit: "
    )
    while True:
        raw = input(prompt_text).strip().lower()
        if raw in _RESUME_CHOICES:
            return _RESUME_CHOICES[raw]
        # Invalid input: loop and re-prompt

File: go/internal/state/state.go
// Package state implements the JSONL state store and resume prompt.
package state

import (
	"bufio"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
	"time"
)

const (
	StatusPushed     = "pushed"
	StatusPROpened   = "pr_opened"
	StatusCommitOnly = "commit_only"
	StatusSkipped    = "skipped"
	StatusFailed     = "failed"
)

var processedStatuses = map[string]bool{
	StatusPushed: true, StatusPROpened: true, StatusCommitOnly: true,
}

// IsProcessed reports whether status counts as processed.
func IsProcessed(status string) bool { return processedStatuses[status] }

var ghUserCharRe = regexp.MustCompile(`^[A-Za-z0-9-]+$`)

// ValidateGHUser returns an error if name is not a valid GitHub username.
// Rules: 1-39 alphanumeric/hyphen chars; no leading/trailing hyphen; no consecutive hyphens.
func ValidateGHUser(name string) error {
	if name == "" || len(name) > 39 || !ghUserCharRe.MatchString(name) {
		return fmt.Errorf("invalid GitHub username: %q", name)
	}
	if name[0] == '-' || name[len(name)-1] == '-' || strings.Contains(name, "--") {
		return fmt.Errorf("invalid GitHub username: %q", name)
	}
	return nil
}

// Clock provides the current time.
type Clock interface{ Now() time.Time }

type realClock struct{}

func (realClock) Now() time.Time { return time.Now().UTC() }

// RealClock returns the production clock (UTC time.Now()).
func RealClock() Clock { return realClock{} }

// FixedClock returns a Clock that always reports ts.
func FixedClock(ts time.Time) Clock { return fixedClock{ts: ts} }

type fixedClock struct{ ts time.Time }

func (f fixedClock) Now() time.Time { return f.ts }

// Record is one entry in the JSONL state file.
type Record struct {
	Repo   string `json:"repo"`
	Status string `json:"status"`
	TS     string `json:"ts"`
	Mode   string `json:"mode,omitempty"`
	Error  string `json:"error,omitempty"`
	PRURL  string `json:"pr_url,omitempty"`
}

// RecordOpts holds optional fields.
type RecordOpts struct {
	Mode  string
	Error string
	PRURL string
}

// Summary aggregates state file content.
type Summary struct {
	Counts      map[string]int
	PRURLs      []string
	FailedRepos []string
}

// Store is an append-only JSONL state file.
type Store struct {
	user     string
	stateDir string
	file     string
	clock    Clock
	mu       sync.Mutex
}

// New constructs a Store. State directory is NOT created until first write.
func New(user, stateDir string, clock Clock) (*Store, error) {
	if err := ValidateGHUser(user); err != nil {
		return nil, err
	}
	if clock == nil {
		clock = RealClock()
	}
	return &Store{
		user:     user,
		stateDir: stateDir,
		file:     filepath.Join(stateDir, "state-"+user+".jsonl"),
		clock:    clock,
	}, nil
}

// File returns the absolute state-file path.
func (s *Store) File() string { return s.file }

// HasPriorState reports whether the state file exists.
func (s *Store) HasPriorState() bool {
	_, err := os.Stat(s.file)
	return err == nil
}

// Record appends one record.
func (s *Store) Record(repo, status string, opts RecordOpts) error {
	rec := Record{
		Repo:   repo,
		Status: status,
		TS:     s.clock.Now().UTC().Format("2006-01-02T15:04:05-07:00"),
		Mode:   opts.Mode,
		Error:  opts.Error,
		PRURL:  opts.PRURL,
	}
	data, err := json.Marshal(rec)
	if err != nil {
		return fmt.Errorf("marshal record: %w", err)
	}
	if err := os.MkdirAll(s.stateDir, 0o755); err != nil {
		return fmt.Errorf("mkdir state dir: %w", err)
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	f, err := os.OpenFile(s.file, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return fmt.Errorf("open state file: %w", err)
	}
	defer f.Close()
	if _, err := f.Write(append(data, '\n')); err != nil {
		return fmt.Errorf("write state: %w", err)
	}
	return nil
}

func (s *Store) readAll() ([]Record, error) {
	f, err := os.Open(s.file)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, fmt.Errorf("open state file: %w", err)
	}
	defer f.Close()
	var out []Record
	sc := bufio.NewScanner(f)
	sc.Buffer(make([]byte, 64*1024), 1024*1024)
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if line == "" {
			continue
		}
		var r Record
		if err := json.Unmarshal([]byte(line), &r); err != nil {
			continue
		}
		out = append(out, r)
	}
	return out, sc.Err()
}

// LoadProcessed returns last-record-wins map of processed repos.
func (s *Store) LoadProcessed() (map[string]Record, error) {
	recs, err := s.readAll()
	if err != nil {
		return nil, err
	}
	last := map[string]Record{}
	for _, r := range recs {
		last[r.Repo] = r
	}
	out := map[string]Record{}
	for repo, r := range last {
		if IsProcessed(r.Status) {
			out[repo] = r
		}
	}
	return out, nil
}

// Summary aggregates counts plus PR URLs and failed repos.
func (s *Store) Summary() (Summary, error) {
	recs, err := s.readAll()
	if err != nil {
		return Summary{}, err
	}
	sum := Summary{Counts: map[string]int{}}
	for _, r := range recs {
		sum.Counts[r.Status]++
		if r.Status == StatusPROpened && r.PRURL != "" {
			sum.PRURLs = append(sum.PRURLs, r.PRURL)
		}
		if r.Status == StatusFailed {
			sum.FailedRepos = append(sum.FailedRepos, r.Repo)
		}
	}
	return sum, nil
}

// ResumeChoice from PromptResume.
type ResumeChoice string

const (
	ResumeKeep  ResumeChoice = "resume"
	ResumeAll   ResumeChoice = "all"
	ResumeFresh ResumeChoice = "fresh"
	ResumeQuit  ResumeChoice = "quit"
)

// PromptResume reads from stdin until a valid choice (r/a/s/q).
func PromptResume(stdin io.Reader, stdout io.Writer, processedCount int) (ResumeChoice, error) {
	r := bufio.NewReader(stdin)
	for {
		fmt.Fprintf(stdout, "Found %d repos already processed. [r]esume (skip processed) / [a]ll incl. failed / [s]tart fresh / [q]uit: ", processedCount)
		line, err := r.ReadString('\n')
		if err != nil && line == "" {
			return "", fmt.Errorf("read stdin: %w", err)
		}
		switch strings.ToLower(strings.TrimSpace(line)) {
		case "r":
			return ResumeKeep, nil
		case "a":
			return ResumeAll, nil
		case "s":
			return ResumeFresh, nil
		case "q":
			return ResumeQuit, nil
		}
	}
}
